jetty-core/jetty-osgi/src/main/java/org/eclipse/jetty/osgi/OSGiDeployableBundleMetadata.java

```diff
@@ -115,139 +130,90 @@ private static String getContextPath(Bundle bundle)
-            {
-                getProperties().put(key, val);
-            }
-            else if (OSGiWebappConstants.JETTY_CONTEXT_FILE_PATH.equalsIgnoreCase(key))
-            {
-                getProperties().put(key, val);
-            }
-        }
+            Dictionary<String, String> properties = new Hashtable<>();
+            properties.put(OSGiWebappConstants.WATERMARK, OSGiWebappConstants.WATERMARK);
 
-        //set up the context path based on the supplied value, or the calculated default
-        setContextPath(getContextPath(bundle));
-    }
+            String bundleSymbolicName = getBundleSymbolicName(contextHandler);
+            if (StringUtil.isNotBlank(bundleSymbolicName))
+                properties.put(OSGiWebappConstants.OSGI_WEB_SYMBOLICNAME, bundleSymbolicName);
 
-    public Resource getBundleResource()
-    {
-        return _bundleResource;
-    }
+            String bundleVersion = getBundleVersionAsString(contextHandler);
+            if (StringUtil.isNotBlank(bundleVersion))
+                properties.put(OSGiWebappConstants.OSGI_WEB_VERSION, bundleVersion);
 
-    @Override
-    public ContextHandler getContextHandler() throws Exception
-    {
-        if (_contextHandler == null)
-                _contextHandler = getAppProvider().createContextHandler(this);
-            return _contextHandler;
-    }
-
-    public void setContextHandler(ContextHandler contextHandler)
-    {
-        _contextHandler = contextHandler;
-    }
+            properties.put(OSGiWebappConstants.OSGI_WEB_CONTEXTPATH, contextHandler.getContextPath());
 
-    public String getPathToResourceBase()
-    {
-        return _pathToResourceBase;
+            serviceRegistration = FrameworkUtil.getBundle(OSGiDeployableBundleMetadata.class).getBundleContext().registerService(ContextHandler.class.getName(), contextHandler, properties);
+            contextHandler.setAttribute(REGISTRATION, serviceRegistration);
+        }
     }
 
-    public void setPathToResourceBase(String path)
+    public static void setBundle(ContextHandler contextHandler, Bundle bundle)
     {
-        _pathToResourceBase = path;
+        contextHandler.setAttribute(BUNDLE, bundle);
     }
 
-    @Override
-    public String getContextPath()
+    public Bundle getBundle()
     {
-        return _contextPath;
+        return bundle;
     }
 
-    public void setContextPath(String contextPath)
+    /**
+     * Convert a bundle installed location into a Resource, taking account of
+     * any locations that are actually packed jars, but without a ".jar" extension, eg
+     * as found on equinox. Eg file:///a/b/c/org.eclipse.osgi/89/0/bundleFile
+     *
+     * @param resourceFactory the ResourceFactory to create Resource from
+     * @return a Resource representing the bundle's installed location
+     */
+    public Resource getBundleResource(ResourceFactory resourceFactory) throws Exception
```

Comment:
This can be a static method, passing in the bundle

Reply:
ok

Reply:
Done